File: Old/KNN.py
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras.datasets import mnist
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
import seaborn as sns
import time

# Load MNIST dataset
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Flatten images for KNN (28x28 -> 784)
train_images_flat = train_images.reshape(train_images.shape[0], -1) / 255.0
test_images_flat = test_images.reshape(test_images.shape[0], -1) / 255.0

# Define k values for small, medium, and high complexity
k_values = {"small": 1, "medium": 5, "high": 15}

for complexity, k in k_values.items():
    print(f"\n=== Running KNN with k={k} ({complexity} complexity) ===")

    # Initialize KNN with the specific k value
    model_knn = KNeighborsClassifier(n_neighbors=k, n_jobs=-1)

    # Measure training time
    print(f"Training KNN with k={k}...")
    start_time = time.time()
    model_knn.fit(train_images_flat, train_labels)
    training_time = time.time() - start_time
    print(f"Training completed in {training_time:.2f} seconds.")

    # Measure inference time
    print(f"Running predictions with k={k}...")
    start_time = time.time()
    knn_predictions = model_knn.predict(test_images_flat)
    inference_time = time.time() - start_time
    print(f"Inference completed in {inference_time:.2f} seconds.")

    # Evaluate accuracy
    knn_accuracy = accuracy_score(test_labels, knn_predictions)
    print(f"KNN Accuracy (k={k}): {knn_accuracy * 100:.2f}%")

    # Confusion Matrix
    cm = confusion_matrix(test_labels, knn_predictions)
    plt.figure(figsize=(10, 8))
    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', cbar=False)
    plt.xlabel('Predicted')
    plt.ylabel('True')
    plt.title(f'Confusion Matrix for KNN (k={k}, {complexity} complexity)')
    plt.show()

    # Classification Report
    print(f"Classification Report (k={k}):")
    print(classification_report(test_labels, knn_predictions))

    # Visualize misclassified digits for each class
    misclassified = np.where(test_labels != knn_predictions)[0]
    misclassified_by_class = {i: [] for i in range(10)}

    for idx in misclassified:
        true_label = test_labels[idx]
        misclassified_by_class[true_label].append(idx)

    plt.figure(figsize=(15, 8))
    for i in range(10):
        if misclassified_by_class[i]:
            idx = misclassified_by_class[i][0]
            plt.subplot(2, 5, i + 1)
            plt.imshow(test_images[idx].reshape(28, 28), cmap='gray')
            plt.title(f"True: {test_labels[idx]}\nPred: {knn_predictions[idx]}")
        else:
            plt.subplot(2, 5, i + 1)
            plt.text(0.5, 0.5, f"No misclassified\nfor {i}", ha='center', va='center')
            plt.axis('off')
    plt.suptitle(f'Misclassified Digits by KNN (k={k}, {complexity} complexity)')
    plt.show()

    # Calculate misclassified counts by digit
    misclassified_counts = {i: 0 for i in range(10)}
    for idx in misclassified:
        true_label = test_labels[idx]
        misclassified_counts[true_label] += 1

    # Calculate successful classification percentage by digit
    success_percentage = {i: 0 for i in range(10)}
    for i in range(10):
        total_samples = np.sum(test_labels == i)
        correct_samples = total_samples - misclassified_counts[i]
        success_percentage[i] = (correct_samples / total_samples) * 100

    # Print results
    print(f"KNN Model Training Time (k={k}): {training_time:.2f} seconds")
    print(f"KNN Model Testing Time (k={k}): {inference_time:.2f} seconds")
    print(f"KNN Model Test Accuracy (k={k}): {knn_accuracy * 100:.2f}%")
    print(f"\nMisclassified counts by digit for KNN (k={k}):")
    for i in range(10):
        print(f"Digit {i}: {misclassified_counts[i]}")
    print(f"\nSuccessful classification percentage for KNN by digit (k={k}):")
    for i in range(10):
        print(f"Digit {i}: {success_percentage[i]:.2f}%")

File: Old/Outputs/KNN/KNN.rs
=== Running KNN with k=1 (small complexity) ===
Training KNN with k=1...
Training completed in 0.04 seconds.
Running predictions with k=1...
Inference completed in 9.15 seconds.
KNN Accuracy (k=1): 96.91%
2025-01-28 13:05:33.672 Python[77561:3706970] +[IMKClient subclass]: chose IMKClient_Modern
2025-01-28 13:05:33.673 Python[77561:3706970] +[IMKInputSession subclass]: chose IMKInputSession_Modern
Classification Report (k=1):
              precision    recall  f1-score   support

           0       0.98      0.99      0.99       980
           1       0.97      0.99      0.98      1135
           2       0.98      0.96      0.97      1032
           3       0.96      0.96      0.96      1010
           4       0.97      0.96      0.97       982
           5       0.95      0.96      0.96       892
           6       0.98      0.99      0.98       958
           7       0.96      0.96      0.96      1028
           8       0.98      0.94      0.96       974
           9       0.96      0.96      0.96      1009

    accuracy                           0.97     10000
   macro avg       0.97      0.97      0.97     10000
weighted avg       0.97      0.97      0.97     10000

KNN Model Training Time (k=1): 0.04 seconds
KNN Model Testing Time (k=1): 9.15 seconds
KNN Model Test Accuracy (k=1): 96.91%

Misclassified counts by digit for KNN (k=1):
Digit 0: 7
Digit 1: 6
Digit 2: 40
Digit 3: 40
Digit 4: 38
Digit 5: 32
Digit 6: 14
Digit 7: 36
Digit 8: 54
Digit 9: 42

Successful classification percentage for KNN by digit (k=1):
Digit 0: 99.29%
Digit 1: 99.47%
Digit 2: 96.12%
Digit 3: 96.04%
Digit 4: 96.13%
Digit 5: 96.41%
Digit 6: 98.54%
Digit 7: 96.50%
Digit 8: 94.46%
Digit 9: 95.84%

=== Running KNN with k=5 (medium complexity) ===
Training KNN with k=5...
Training completed in 0.03 seconds.
Running predictions with k=5...
Inference completed in 10.71 seconds.
KNN Accuracy (k=5): 96.88%
Classification Report (k=5):
              precision    recall  f1-score   support

           0       0.96      0.99      0.98       980
           1       0.95      1.00      0.98      1135
           2       0.98      0.96      0.97      1032
           3       0.96      0.97      0.97      1010
           4       0.98      0.96      0.97       982
           5       0.97      0.97      0.97       892
           6       0.98      0.99      0.98       958
           7       0.96      0.96      0.96      1028
           8       0.99      0.94      0.96       974
           9       0.96      0.95      0.95      1009

    accuracy                           0.97     10000
   macro avg       0.97      0.97      0.97     10000
weighted avg       0.97      0.97      0.97     10000

KNN Model Training Time (k=5): 0.03 seconds
KNN Model Testing Time (k=5): 10.71 seconds
KNN Model Test Accuracy (k=5): 96.88%

Misclassified counts by digit for KNN (k=5):
Digit 0: 6
Digit 1: 2
Digit 2: 41
Digit 3: 34
Digit 4: 38
Digit 5: 30
Digit 6: 13
Digit 7: 40
Digit 8: 61
Digit 9: 47

Successful classification percentage for KNN by digit (k=5):
Digit 0: 99.39%
Digit 1: 99.82%
Digit 2: 96.03%
Digit 3: 96.63%
Digit 4: 96.13%
Digit 5: 96.64%
Digit 6: 98.64%
Digit 7: 96.11%
Digit 8: 93.74%
Digit 9: 95.34%

=== Running KNN with k=15 (high complexity) ===
Training KNN with k=15...
Training completed in 0.04 seconds.
Running predictions with k=15...
Inference completed in 10.73 seconds.
KNN Accuracy (k=15): 96.33%
Classification Report (k=15):
              precision    recall  f1-score   support

           0       0.96      0.99      0.98       980
           1       0.94      1.00      0.97      1135
           2       0.98      0.94      0.96      1032
           3       0.96      0.97      0.96      1010
           4       0.97      0.95      0.96       982
           5       0.97      0.97      0.97       892
           6       0.97      0.98      0.98       958
           7       0.95      0.95      0.95      1028
           8       0.99      0.93      0.96       974
           9       0.94      0.95      0.95      1009

    accuracy                           0.96     10000
   macro avg       0.96      0.96      0.96     10000
weighted avg       0.96      0.96      0.96     10000

KNN Model Training Time (k=15): 0.04 seconds
KNN Model Testing Time (k=15): 10.73 seconds
KNN Model Test Accuracy (k=15): 96.33%

Misclassified counts by digit for KNN (k=15):
Digit 0: 10
Digit 1: 4
Digit 2: 64
Digit 3: 35
Digit 4: 48
Digit 5: 29
Digit 6: 15
Digit 7: 48
Digit 8: 67
Digit 9: 47

Successful classification percentage for KNN by digit (k=15):
Digit 0: 98.98%
Digit 1: 99.65%
Digit 2: 93.80%
Digit 3: 96.53%
Digit 4: 95.11%
Digit 5: 96.75%
Digit 6: 98.43%
Digit 7: 95.33%
Digit 8: 93.12%
Digit 9: 95.34%